### DISCUSSION TITLE
Optimized example | Dynamic components inquiry

### DISCUSSION THREAD

**Discussion by iamshz97** [QuestPDF/QuestPDF#358] [Q&A] [answered]
Hi Team,

I'm new to QuestPDF and loving it I'm currently playing with the example in the documentation which shows us an Optimized example on dealing with Dynamic components.

![image](https://user-images.githubusercontent.com/33499387/194150768-c29b878c-360d-4882-92f4-2688990ad821.png)

I did follow the same and I getting the following error 

![image](https://user-images.githubusercontent.com/33499387/194150432-ea2ab252-e73d-4ecd-b321-6b813262267c.png)

Any quick workarounds or am I missing something here...
Much Appreciated 😍 Quest PDF is a masterpiece.

**Comment by MarcinZiabek** [comment, marked as answer]
Thank you for your kind words 😁 

I will take a look if this API is actually public. In the test project, I am using the InternalsVisibleTo capability. That means, I might have forgot to make this methods publicly available. Will do my best to do within next couple of days.

In the meantime, would you like to take a look at this PR and share your thoughs? https://github.com/QuestPDF/QuestPDF/pull/357

**Reply by iamshz97** [reply]
Amazing @MarcinZiabek much thanks for the quick response looking forward to it, sorry for the delay from my end I'm from another zone.

I had a look in the PR and that's a quite interesting feature, like you've mentioned this is definitely something that can cater a use-case. One of those things when you're like Oh this is what I need.

In terms of feedback everything looks perfect to me this will be definitely be a value addition ✨

**Reply by iamshz97** [reply]
@MarcinZiabek I was also trying to see is there a way that I can get the **current page number** and **height left before page break** while iterating and creating rows in a table and then do a page break but I assume there's limitation on achieving that ? and that's why we are following the use of this approach(Dynamic components).

I had couple of things in mind without following this approach, but definitely not the best way but I'll share my thoughts

Requirement: At the first row of each page color the first cell..

1. Have a variable to know how many items can be generated per page and then iterate and do
```Csharp

       // for every 35 row items on list
        container
            .Page(page =>
            {
                page.Size(PageSizes.A4);
                page.Margin(50);
                    
                page.Header().Element(ComposeHeader);
                page.Content().Element(ComposeContent);
                page.Footer().Element(ComposeFooter);
            });
```

This way I can figure out I'm creating the first row on a new page with the help of additional properties

limitation following this is I have to make sure each row has a fixed height and cannot change based on content, which is what our old 3rd part report generating app does (I know sounds odd).

The Table with per-page totals approach did help but the performance is noticeable but still it is still fine. But I just wanted to figure out is there a better way....

Either way once the API is public I should be able to achieve what I need.
Many Thanks.

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
